rfcs/0001-messages.md

```diff
@@ -0,0 +1,414 @@
+# FROST messages
+
+Proposes a message layout to exchange information between participants of a FROST setup using the [jubjub](https://github.com/zkcrypto/jubjub) curve.
+
+## Motivation
+
+Currently FROST library is complete for 2 round signatures with a dealer/aggregator setup.
+This proposal acknowledges that specific features, additions and upgrades will need to be made when DKG is implemented.
+
+Assuming all participants have a FROST library available, we need to define message structures in a way that data can be exchanged between participants. The proposal is a collection of data types so each side can do all the actions needed for a real life situation.
+
+## Definitions
+
+- `dealer` - Participant who distributes the initial package to all the other participants.
+- `aggregator` - Participant in charge of collecting all the signatures from the other participants and generating the final group signature.
+- `signer` - Participant that will receive the initial package, sign and send the signature to the aggregator to receive the final group signature.
+
+Note: In this RFC we consider the above 3 participants to be different. `dealer` and `aggregator` have specific hard coded `ParticipantId`s, so for example a `dealer` can't be a `signer`. This is not a protocol limitation but a specific rule introduced in this document.
+
+## Guide-level explanation
+
+We propose a message separated in 2 parts, a header and a payload:
+
+```rust
+/// The data required to serialize a frost message.
+struct Message {
+    header: Header,
+    payload: Payload,
+}
+```
+
+`Header` will look as follows:
+
+```rust
+/// The data required to serialize the common header fields for every message.
+///
+/// Note: the `msg_type` is derived from the `payload` enum variant.
+struct Header {
+    version: MsgVersion,
+    sender: ParticipantId,
+    receiver: ParticipantId,
+}
+```
+
+While `Payload` will be defined as:
+
+```rust
+/// The data required to serialize the payload for a message.
+enum Payload {
+    SharePackage(messages::SharePackage),
+    SigningCommitments(messages::SigningCommitments),
+    SigningPackage(messages::SigningPackage),
+    SignatureShare(messages::SignatureShare),
+    AggregateSignature(messages::AggregateSignature),
+}
+```
+
+All the messages and new types will be defined in a new file `src/frost/messages.rs`
+
+## Reference-level explanation
+
+Here we explore in detail the header types and all the message payloads. 
+
+### Header
+
+Fields of the header define new types. Proposed implementation for them is as follows:
+
+```rust
+/// The numeric values used to identify each `Payload` variant during serialization.
+#[repr(u8)]
+#[non_exhaustive]
+enum MsgType {
+    SharePackage,
+    SigningCommitments,
+    SigningPackage,
+    SignatureShare,
+    AggregateSignature,
+}
+
+/// The numeric values used to identify the protocol version during serialization.
+struct MsgVersion(u8);
```

Comment:
@teor2345 @oxarbitrage what if this is an enum instead, also `repr(u8)`? So that there will only be a strict set of supported versions. Yes the first pass at this will be just an enum with a single variant, the current default, which serializes to 0u8 or whatever.

Reply:
I think an enum is a good idea. But that change isn't urgent, because it doesn't change the wire format.

Can you open a ticket to update the implementation after Zcon?

Reply:
Thanks, this is now https://github.com/ZcashFoundation/redjubjub/issues/112